File: ubag-python/src/ubag/_keys.py
"""
Asymmetric key primitives for UBAG.

Two key types, each chosen for a reason:

* Agent identity  -> Ed25519. Raw 64-byte signatures, no DER/encoding ambiguity,
  so an agent signature produced in Python verifies byte-for-byte in Node and
  vice-versa. The agent's PUBLIC key is its identity.

* Issuer (credential signing) -> EC P-256 / ES256 JWT. Chosen because it is the
  one asymmetric JWT algorithm supported by BOTH PyJWT and Node's `jsonwebtoken`,
  so a credential minted by either SDK verifies with the other via the issuer's
  PUBLIC key (JWKS) — the OAuth/OIDC model. No shared secret, works across sites.
"""
from __future__ import annotations

import base64
import hashlib

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec, ed25519


def _b64u(b: bytes) -> str:
    return base64.urlsafe_b64encode(b).decode("ascii").rstrip("=")


def _b64u_decode(s: str) -> bytes:
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


# ── Agent identity keys (Ed25519) ──────────────────────────────────────────────

def generate_agent_keypair() -> tuple[str, str]:
    """Generate an agent identity keypair. Returns (private_b64url, public_b64url)
    of the raw 32-byte Ed25519 keys."""
    sk = ed25519.Ed25519PrivateKey.generate()
    priv = sk.private_bytes(
        serialization.Encoding.Raw,
        serialization.PrivateFormat.Raw,
        serialization.NoEncryption(),
    )
    pub = sk.public_key().public_bytes(
        serialization.Encoding.Raw, serialization.PublicFormat.Raw
    )
    return _b64u(priv), _b64u(pub)


def agent_sign(private_b64: str, message: bytes) -> str:
    """Sign a message with an agent's Ed25519 private key. Returns b64url signature."""
    sk = ed25519.Ed25519PrivateKey.from_private_bytes(_b64u_decode(private_b64))
    return _b64u(sk.sign(message))


def agent_verify(public_b64: str, message: bytes, signature_b64: str) -> bool:
    """Verify an agent's Ed25519 signature over `message`. Never raises."""
    try:
        pk = ed25519.Ed25519PublicKey.from_public_bytes(_b64u_decode(public_b64))
        pk.verify(_b64u_decode(signature_b64), message)
        return True
    except Exception:
        return False


def agent_id(public_b64: str) -> str:
    """Stable agent identity derived from its public key (a key thumbprint).
    This is what `sub` / `cnf` reference — identity is the key, not a claimed name."""
    digest = hashlib.sha256(_b64u_decode(public_b64)).digest()
    return "ubag:" + _b64u(digest)[:43]


# ── Issuer keys (EC P-256, for ES256 credential JWTs) ──────────────────────────

def generate_issuer_keypair() -> tuple[str, str]:
    """Generate an issuer keypair for signing credentials.
    Returns (private_pem, public_pem) for an EC P-256 key (ES256)."""
    sk = ec.generate_private_key(ec.SECP256R1())
    priv = sk.private_bytes(
        serialization.Encoding.PEM,
        serialization.PrivateFormat.PKCS8,
        serialization.NoEncryption(),
    ).decode("ascii")
    pub = sk.public_key().public_bytes(
        serialization.Encoding.PEM,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode("ascii")
    return priv, pub


def issuer_public_from_private(private_pem: str) -> str:
    """Derive the issuer public PEM from its private PEM (so a site configured only
    with the private key can also verify)."""
    sk = serialization.load_pem_private_key(private_pem.encode("ascii"), password=None)
    return sk.public_key().public_bytes(
        serialization.Encoding.PEM,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode("ascii")


def build_jwks(public_pem: str, kid: str = "ubag-issuer-1") -> dict:
    """Build a JWKS document for the issuer's EC P-256 public key, so resource
    servers can verify credentials by fetching /.well-known/jwks.json."""
    pk = serialization.load_pem_public_key(public_pem.encode("ascii"))
    nums = pk.public_numbers()
    x = nums.x.to_bytes(32, "big")
    y = nums.y.to_bytes(32, "big")
    return {
        "keys": [
            {
                "kty": "EC",
                "crv": "P-256",
                "x": _b64u(x),
                "y": _b64u(y),
                "use": "sig",
                "alg": "ES256",
                "kid": kid,
            }
        ]
    }

File: ubag-node/src/keys.js
'use strict';

/**
 * Asymmetric key primitives for UBAG (mirrors ubag-python/src/ubag/_keys.py).
 *
 *  - Agent identity  -> Ed25519, serialized as RAW base64url keys so a signature
 *    produced here verifies byte-for-byte in the Python SDK and vice-versa.
 *  - Issuer (credential signing) -> EC P-256 / ES256 JWT (PEM), the one asymmetric
 *    JWT algorithm both `jsonwebtoken` and PyJWT support, so credentials are
 *    cross-SDK verifiable via the issuer PUBLIC key (JWKS) — the OAuth/OIDC model.
 */
const crypto = require('crypto');

// Fixed DER wrappers so we can round-trip RAW Ed25519 keys through Node's KeyObject API.
const ED25519_SPKI_PREFIX = Buffer.from('302a300506032b6570032100', 'hex'); // 12 bytes
const ED25519_PKCS8_PREFIX = Buffer.from('302e020100300506032b657004220420', 'hex'); // 16 bytes

const b64u = (buf) => Buffer.from(buf).toString('base64url');
const unb64u = (s) => Buffer.from(s, 'base64url');

// ── Agent identity (Ed25519, raw b64url) ──────────────────────────────────────
function generateAgentKeypair() {
  const { publicKey, privateKey } = crypto.generateKeyPairSync('ed25519');
  const pubDer = publicKey.export({ format: 'der', type: 'spki' });
  const privDer = privateKey.export({ format: 'der', type: 'pkcs8' });
  return { privateKey: b64u(privDer.subarray(-32)), publicKey: b64u(pubDer.subarray(-32)) };
}

function _privKeyObj(privB64) {
  return crypto.createPrivateKey({
    key: Buffer.concat([ED25519_PKCS8_PREFIX, unb64u(privB64)]),
    format: 'der',
    type: 'pkcs8',
  });
}
function _pubKeyObj(pubB64) {
  return crypto.createPublicKey({
    key: Buffer.concat([ED25519_SPKI_PREFIX, unb64u(pubB64)]),
    format: 'der',
    type: 'spki',
  });
}

function agentSign(privB64, message) {
  return b64u(crypto.sign(null, Buffer.from(message), _privKeyObj(privB64)));
}
function agentVerify(pubB64, message, signatureB64) {
  try {
    return crypto.verify(null, Buffer.from(message), _pubKeyObj(pubB64), unb64u(signatureB64));
  } catch {
    return false;
  }
}
function agentId(pubB64) {
  const digest = crypto.createHash('sha256').update(unb64u(pubB64)).digest('base64url');
  return 'ubag:' + digest.slice(0, 43);
}

// ── Issuer keys (EC P-256, for ES256 credential JWTs) ─────────────────────────
function generateIssuerKeypair() {
  const { publicKey, privateKey } = crypto.generateKeyPairSync('ec', {
    namedCurve: 'P-256',
    publicKeyEncoding: { type: 'spki', format: 'pem' },
    privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
  });
  return { privateKey, publicKey };
}
function issuerPublicFromPrivate(privPem) {
  return crypto.createPublicKey(privPem).export({ type: 'spki', format: 'pem' });
}
function buildJwks(pubPem, kid = 'ubag-issuer-1') {
  const jwk = crypto.createPublicKey(pubPem).export({ format: 'jwk' });
  return { keys: [{ ...jwk, use: 'sig', alg: 'ES256', kid }] };
}

module.exports = {
  b64u,
  unb64u,
  generateAgentKeypair,
  agentSign,
  agentVerify,
  agentId,
  generateIssuerKeypair,
  issuerPublicFromPrivate,
  buildJwks,
};
